### DISCUSSION TITLE
Using `custom` with a predefined schema?

### DISCUSSION THREAD

**Discussion by chrislgrigg** [open-circle/valibot#508] [Q&A] [answered]
Hello! I'm exploring a migration from Zod to Valibot and stuck on what I assume is a simple issue.

In Zod, I can take a schema, `fooSchema`, and use `refine` later on to add additional logic, perhaps in a form that has some unique quality that shouldn't be shared in the base schema definition. Valibot's equivalent seems to be `custom`, but I can only use this in a pipeline when defining the original schema. I'm looking for something like this:

```
const fooSchema = v.object({ foo: v.string([v.toTrimmed(), v.email()]) });

// This does not work, what is the correct equivalent?
const extendedFoo = v.object(fooSchema, v.custom(({ foo }) => {
  // some additional logic here
});
```

Thank you for your help!

**Comment by fabian-hiller** [comment, marked as answer]
This is already possible, but I am aware that it is a bit cumbersome. That's why we're currently changing that part of the API in #502.

> Check the code out in our [playground](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgDEIIBlJgBYBTEMjgBeTADoIFAFZCmMABQBvHDwBc0jlGCsA5soDaGKTAgAVPSBBCAJsoCUAGmkjkwUs4C6TuAC+TgzMbBxwAKIAHjBCrPYO3BAS0rIKSspJ-MKiUnEwekKobsb0cNJMAK4c+MpqGslBEgB8DVL6TKSVCajKhAACBqJehK5EEVB4UNoActQocADiw6QAhKP0fiEs7PBQRZWk8JJmqMjYQgAKyFCoQsrRsfGJPG6qZQ3ahHLIrEL9QiiyHApCEUhYIEILnoQW2YQgoKkpAgRn2qEOMGCQA)

```ts
import * as v from 'valibot';

const FooSchema = v.object({ foo: v.string([v.toTrimmed(), v.email()]) });

const ExtendedFoo = v.object(FooSchema.entries, [
  v.custom(({ foo }) => foo.includes('@gmail'), 'Error: Not a Gmail!')
]);
```

**Reply by fabian-hiller** [reply]
In the future, you can write the following:

```ts
import * as v from 'valibot';

const FooSchema = v.object({ foo: v.pipe(v.string(), v.toTrimmed(), v.email()) });

const ExtendedFoo = v.pipe(
  FooSchema,
  v.check(({ foo }) => foo.includes('@gmail'), 'Error: Not a Gmail!')
);
```

**Comment by chrislgrigg** [comment]
This is great! The current syntax isn’t terrible but the new syntax is quite exciting and will be a very significant improvement. Thank you so much for the quick reply and all your work!
